src/protocol/attribution/aggregate_credit.rs

```diff
@@ -0,0 +1,358 @@
+
+    async fn reshare<C>(&self, ctx: C, record_id: RecordId, to_helper: Role) -> Result<Self, Error>
+    where
+        C: Context<F, Share = <Self as Resharable<F>>::Share> + Send,
+    {
+        let f_helper_bit = ctx
+            .narrow(&HelperBit)
+            .reshare(&self.helper_bit, record_id, to_helper);
+        let f_aggregation_bit =
+            ctx.narrow(&AggregationBit)
+                .reshare(&self.aggregation_bit, record_id, to_helper);
+        let f_breakdown_key =
+            ctx.narrow(&BreakdownKey)
+                .reshare(&self.breakdown_key, record_id, to_helper);
+        let f_value = ctx
+            .narrow(&Credit)
+            .reshare(&self.credit, record_id, to_helper);
+
+        let mut outputs =
+            try_join_all([f_helper_bit, f_aggregation_bit, f_breakdown_key, f_value]).await?;
+
+        Ok(CappedCreditsWithAggregationBit {
+            helper_bit: outputs.remove(0),
+            aggregation_bit: outputs.remove(0),
+            breakdown_key: outputs.remove(0),
+            credit: outputs.remove(0),
+        })
+    }
+}
+
+/// Transpose rows of bits into bits of rows
+///
+/// input:
+/// [
+///   [ row[0].bit0, row[0].bit1, ..., row[0].bit31 ],
+///   [ row[1].bit0, row[1].bit1, ..., row[1].bit31 ],
+///   ...
+///   [ row[n].bit0, row[n].bit1, ..., row[n].bit31 ],
+/// ]
+///
+/// output:
+/// [
+///   [ row[0].bit0,  row[1].bit0,  ..., row[n].bit0 ],
+///   [ row[0].bit1,  row[1].bit1,  ..., row[n].bit1 ],
+///   ...
+///   [ row[0].bit31, row[1].bit31, ..., row[n].bit31 ],
+/// ]
+fn transpose<F: Field>(input: &[Vec<Replicated<F>>]) -> Vec<Vec<Replicated<F>>> {
+    let bit_length = input[0].len();
+    debug_assert_eq!(bit_length, F::Integer::BITS as usize);
+
+    (0..bit_length)
+        .map(|i| input.iter().map(|b| b[i].clone()).collect::<Vec<_>>())
+        .collect::<Vec<_>>()
+}
+
+async fn bit_decompose_breakdown_key<F: Field>(
+    ctx: SemiHonestContext<'_, F>,
+    input: &[CappedCreditsWithAggregationBit<F>],
+) -> Result<Vec<Vec<Replicated<F>>>, Error> {
+    let random_bits_generator = RandomBitsGenerator::new();
+    try_join_all(
+        input
+            .iter()
+            .zip(repeat(ctx))
+            .enumerate()
+            .map(|(i, (x, c))| {
+                let rbg = random_bits_generator.clone();
+                async move {
+                    BitDecomposition::execute(c, RecordId::from(i), rbg, &x.breakdown_key).await
+                }
+            })
+            .collect::<Vec<_>>(),
+    )
+    .await
+}
+
+/// Sort the input by `aggregation_bit` first, then by `breakdown_key`
+#[allow(dead_code)]
+async fn sort_by_aggregation_bit_and_breakdown_key<F: Field>(
+    ctx: SemiHonestContext<'_, F>,
+    input: &[CappedCreditsWithAggregationBit<F>],
+) -> Result<Vec<CappedCreditsWithAggregationBit<F>>, Error> {
+    // Sort by aggregation_bit
+    let sorted_by_aggregation_bit = sort_by_aggregation_bit(ctx.clone(), input).await?;
```

Comment:
For the future, I wonder if it is better to just pre-pend the bit-decomposed breakdown keys with this bit and generate a single permutation, which is applied just once.